lambda/matching.ml

```diff
@@ -2817,6 +2802,76 @@ let mk_failaction_pos partial seen ctx defs =
       )
     ;
     (fail, [], jumps)
+  ) else (
+    (* Compute a specialized context for each failure pattern. *)
+    let fail_pats_in_ctx =
+      List.map (fun pat -> (pat, Context.lub pat ctx)) fail_pats in
+    (* Determine which exit is appropriate for which failure patterns. *)
+    let rec exit_failpats defs fail_pats_in_ctx acc =
+      match Default_environment.pop defs with
+      | Some ((idef, pss), rem) ->
+          (* Collect the patterns whose context matches the
+             matrix [pss] of the exit [idef]. *)
+          let now, later =
+            List.partition_map (fun ((p, p_ctx) as fail_pat) ->
+              if Context.matches p_ctx pss
+              then Either.Left p
+              else Either.Right fail_pat
+            ) fail_pats_in_ctx
+          in
+          let acc =
+            match now with
+            | [] -> acc
+            | _ :: _ -> (idef, now) :: acc
+          in
+          begin match later with
+          | _ :: _ -> exit_failpats rem later acc
+          | [] ->
+              (* We have assigned exit point to all fail patterns, so
+                 we can stop iterating on the exits. *)
+              acc
+          end
+      | None ->
+          (* If there are no exits left in the environment, we
+             consider that the remaining failing patterns cannot
+             actually arise. [mk_failaction_neg] has the same
+             logic. *)
+          acc
+    in
+    let exit_failpats = exit_failpats defs fail_pats_in_ctx [] in
+    let fails =
+      List.concat_map (fun (i, i_fail_pats) ->
+        let action = Lstaticraise (i, []) in
+        List.map (fun pat -> (get_key_constr pat, action)) i_fail_pats
+      ) exit_failpats
```

Comment:
I believe this inverts the order of clauses from before, as the outer `fold_left` has been turned into a `concat_map`:

```ocaml
| A | B -> exit 1
| C | D -> exit 2

(* becomes *)

| C | D -> exit 2
| A | B -> exit 1
```

I don't believe this was an intentional or important property of the old code, and the new order should be semantically equivalent, so feel free to resolve this comment.